examples/tf/cluster_demo.py

```diff
@@ -21,30 +22,34 @@ def run_task(v):
     run_experiment can easily execute variants of the experiment on different
     machines
     """
-    env = TfEnv(env_name="CartPole-v1")
+    with LocalRunner() as runner:
```

Comment:
can you explain the use of a context manager here? what are you trying to express?

Reply:
This is to provide a default tensorflow session for env, policy, and algorithm. We are having some mess with which session to use and where to create a session now, so I make the following assumptions. It's the most natrual solution to me.

- A session can only be created by runner.
- A default session is always readily available when policy and algo are constructed and used.
Note: The new `GaussianMLPPolicyWithModel` requires a session during construction, so creating session as we are doing now is not a good idea.
- No where in policy and algo implementation should `tf.global_variables_initializer()` be called. Instead, they only (optionally) initialize variables they own.
- The runner initializes all uninitialized variable after policy and algo are constructed and before the training starts.

Reply:
I like these assumptions, though I want to explore in the future whether a ContextManager is the best way to implement them. I am happy with this for noe

By the way, you could implement your 3rd bullet by monkey-patching `tf.global_variables_initializer()` inside the context manager:
```python
import mock

_TF_GLOBAL_VARIABLES_INITIALIZER_SYMBOLS = [
    'tf.global_variables_initializer',
    'tf.initializers.global_variables',
    'tf.initialize_all_variables',
]

def _complain(symbol):
    raise RuntimeError('garage does not supporting calling {}'.format(symbol))

class TfRunner:
    def __init__(self):

        self._patch_no_globals = [mock.patch(s, _complain(s)) for s in TF_GLOBAL_VARIABLES_INITIALIZER_SYMBOLS]

    def __enter__(self):
        for p in self._patch_no_globals:
            p.__enter__()

    def __exit__(self):
        for p in reversed(self._patch_no_globals):
            p.__exit__(self):


class LocalRunner(TfRunner):
    def __init__(self):
        super().__init__()

    def __enter__(self):
        super().__enter__()
        # stuff

    def __exit__(self):
        # stuff
        super().__exit__()
```

Reply:
Cool! LocalRunner depends on tensorflow now and one thing we need in the future is to support for frameworks other than tensorflow. Your snippet demonstrates a nice shape of doing it.